Add complete documentation of the message retention policies support (#6624)
matrix-org/synapse: docs/message_retention_policies.md

# Message retention policies

> Synapse admins can enable support for message retention policies on

Reviewer: this hasn't yet completed FCP so needs to have a big EXPERIMENTAL stamp on it.

their homeserver. Message retention policies exist at a room level,
follow the semantics described in
[MSC1763](https://github.com/matrix-org/matrix-doc/blob/matthew/msc1763/proposals/1763-configurable-retention-periods.md),
and allow server and room admins to configure how long messages should
be kept in a homeserver's database before being purged from it.

A message retention policy is mainly defined by its `max_lifetime`
parameter, which defines how long a message can be kept around after

> it's been sent in the room. If a room doesn't have a message retention

Reviewer: ```suggestion
it was sent to the room. If a room doesn't have a message retention
```

policy, and there's no default one for a given server, then no message
sent in that room is ever purged on that server.

MSC1763 also specifies semantics for a `min_lifetime` parameter which
defines the amount of time after which an event _can_ get purged (after

> it's been sent to the room), but Synapse doesn't currently support it

Reviewer: ```suggestion
it was sent to the room), but Synapse doesn't currently support it
```

Since purge jobs don't run continuously, this means that an event might
stay in a server's database for longer than the value for `max_lifetime`
in the room would allow, though hidden from clients.

Similarly, if a server (with support for message retention policies
enabled) receives from another server an event that should have been
purged according to its room's policy, then the receiving server will
process and store that event until it's picked up by the next purge job,
though it will always hide it from clients.

## Room configuration

To configure a room's message retention policy, a room's admin or
moderator needs to send a state event in that room with the type
`m.room.retention` and the following content:

```json
{
    "max_lifetime": ...
}
```

In this event's content, the `max_lifetime` parameter has the same
meaning as previously described, and needs to be expressed in
milliseconds. The event's content can also include a `min_lifetime`
parameter, which has the same meaning and limited support as previously
described.

Note that over every server in the room, only the ones with support for
message retention policies will actually remove expired events. While

> we plan to eventually enable this support by default in Synapse, this

Reviewer: this will cause outrage; I'm not sure we've decided to do it yet. suggest just saying it's not enabled by default in synapse.

Author: I've based that assumption on [this comment](https://github.com/matrix-org/matrix-doc/pull/1763#discussion_r317596128), but fair enough.

isn't currently the case.

> ## Server configuration

Reviewer: this should probably be at the top? My first question here is "what will my server do with this stuff".

Support for this feature can be enabled and configured in the
`retention` section of the Synapse configuration file (see the
[sample file](https://github.com/matrix-org/synapse/blob/v1.7.3/docs/sample_config.yaml#L332-L393)).

> To enable support for message retentions policies, set the setting

Reviewer: ```suggestion
To enable support for message retention policies, set the setting
```

`enabled` in this section to `true`.

### Default policy

A default message retention policy is a policy defined in Synapse's
configuration that is used by Synapse for every room that doesn't have a
message retention policy configured in its state. This allows server
admins to ensure that messages are never kept indefinitely in a server's
database. 

A default policy can be defined as such, in the `retention` section of
the configuration file:

```yaml
  default_policy:
    min_lifetime: 1d
    max_lifetime: 1y
```

Here, `min_lifetime` and `max_lifetime` have the same meaning and level
of support as previously described. They can be expressed either as a
duration (using the units `s` (seconds), `m` (minutes), `h` (hours),
`d` (days), `w` (weeks) and `y` (years)) or as a number of milliseconds.

### Purge jobs

> Purge jobs are the jobs that Synapse run in the background to purge

Reviewer: ```suggestion
Purge jobs are the jobs that Synapse runs in the background to purge
```

expired events from the database. They are only run if support for
message retention policies is enabled in the server's configuration. If
no configuration for purge jobs is configured by the server admin,

> Synapse will run one daily that will handle every room with a message

Reviewer: I'm a little bit worried about duplicating this information from the sample_config: I think we might forget to update this if it changes. Can we just say "there is a default, see the sample config for details"?
